src/drivers/net/marvell/aqc1xx.c

```diff
@@ -0,0 +1,634 @@
+#include <compiler.h>
+
+#include "aqc1xx.h"
+
+
+static int atl_ring_alloc(const struct atl_nic *nic, struct atl_ring *ring,
+			  uint32_t desc_size, uint32_t reg_base)
+{
+	physaddr_t phy_addr;
+
+	/* Allocate ring buffer.*/
+	ring->length = ATL_RING_SIZE * desc_size;
+	ring->ring = malloc_phys(ring->length, ATL_RING_ALIGN);
+	if (!ring->ring)
+		goto err_alloc;
+	ring->sw_head = ring->sw_tail = 0;
+
+	memset(ring->ring, 0, ring->length);
+
+	/* Write ring address (hi & low parts).*/
+	phy_addr = virt_to_bus(ring->ring);
+	ATL_WRITE_REG((uint32_t)phy_addr, reg_base + 0);
+	ATL_WRITE_REG((uint32_t)(((uint64_t)phy_addr) >> 32), reg_base + 4);
+	/* Write ring length.*/
+	ATL_WRITE_REG(ATL_RING_SIZE, reg_base + 8);
+
+	/* #todo: reset head and tail pointers */
+	DBG("ATLANTIC %p ring is at [%08llx,%08llx), reg base %#x\n",
+	    nic, ((unsigned long long)phy_addr),
+	    ((unsigned long long) phy_addr + ring->length), reg_base);
+
+	return 0;
+
+err_alloc:
+	if (ring->ring)
+		free_phys(ring->ring, ring->length);
+
+	ring->ring = NULL;
+	ring->length = 0;
+
+	return -ENOMEM;
+}
+
+static void atl_ring_free(struct atl_ring *ring)
+{
+	if (ring->ring)
+		free_phys(ring->ring, ring->length);
+	ring->ring = NULL;
+	ring->length = 0;
+}
+
+static void atl_ring_next_dx(unsigned int *val)
+{
+	++(*val);
+	if (*val == ATL_RING_SIZE)
+		*val = 0;
+}
+
+int atl_ring_full(const struct atl_ring *ring)
+{
+	unsigned int tail = ring->sw_tail;
+	atl_ring_next_dx(&tail);
+	return tail == ring->sw_head;
+}
+
+void atl_rx_ring_fill(struct atl_nic *nic)
+{
+	struct atl_desc_rx *rx;
+	struct io_buffer *iobuf;
+	physaddr_t address;
+	unsigned int refilled = 0;
+
+	/* Refill ring */
+	while (!atl_ring_full(&nic->rx_ring)) {
+
+		/* Allocate I/O buffer */
+		iobuf = alloc_iob(ATL_RX_MAX_LEN);
+		if (!iobuf) {
+			/* Wait for next refill */
+			break;
+		}
+
+		/* Get next receive descriptor */
+		rx = (struct atl_desc_rx *)nic->rx_ring.ring +
+		nic->rx_ring.sw_tail;
+
+		/* Populate receive descriptor */
+		address = virt_to_bus(iobuf->data);
+		rx->data_addr = address;
+		rx->hdr_addr = 0;
+
+		/* Record I/O buffer */
+		assert(nic->iobufs[nic->rx_ring.sw_tail] == NULL);
+		nic->iobufs[nic->rx_ring.sw_tail] = iobuf;
+
+		DBG("AQUANTIA RX[%d] is [%llx,%llx)\n", nic->rx_ring.sw_tail,
+		    ((unsigned long long) address),
+		    ((unsigned long long) address + ATL_RX_MAX_LEN));
+		atl_ring_next_dx(&nic->rx_ring.sw_tail);
+		refilled++;
+	}
+
+	/* Push descriptors to card, if applicable */
+	if (refilled) {
+		wmb();
+		ATL_WRITE_REG(nic->rx_ring.sw_tail, ATL_RING_TAIL_PTR);
+	}
+}
+
+/**
+* Open network device
+*
+* @v netdev		Network device
+* @ret rc		Return status code
+*/
+static int atl_open(struct net_device *netdev)
+{
+	struct atl_nic *nic = netdev->priv;
+	uint32_t ctrl = 0;
+	DBG("AQUANTIA: atl_open()\n");
+
+	/* Tx ring */
+	if (atl_ring_alloc(nic, &nic->tx_ring, sizeof(struct atl_desc_tx),
+	    ATL_TX_DMA_DESC_ADDR) != 0)
+		goto err_alloc;
```

Comment:
Please use separate error handling labels for TX and RX allocation failures.  See `intel_open()` for an example.  This will allow you to eliminate all of the checks for `ring->ring == NULL`.